### PR TITLE
More VM unit tests

### PR DESCRIPTION
**Pull request by ducky64**
DO NOT MERGE - work in progress

More VM example programs, more unit tests for those programs using the JDOM unit test infrastructure, and refactoring the VM test helper utilities.

Summary:
- Simple tests for the remember servo program (set/recall servo position) and two-button servo program (two buttons to set servo to fixed positions).
- (currently broken) test for button-to-switch adapter (click button to toggle switch position)
  - this will be the first VM test using server roles
- add pot to switch program (use pot as a switch using fixed threshold with hysteresis)

### CODE REVIEW

**Comment by tballmsft** [comment]
I added some debug logging. The reason the event is not going through is shown by the following trace:
```
cd tests && node ../node_modules/mocha/bin/mocha --require ts-node/register --watch-extensions ts,tsx -p ./tsconfig.json **/buttontoswitch.spec.ts

start host
new device CP30 (a93f5f5a30a60ea2)
new device KZ18 (644260d120c19416)
start host
new device SY15 (aaecd1ebeadc2043)
starting data=0 unpacked=0
start host
new device WI21 (30349c5382b255af)
new device LC06 (d6ad9352d4e0f561)
start host
new device XJ73 (458c19689f26b244)
sw device = XJ73
bus event at 1000: code=1 device=WI21

  button to switch adapter
    √ switch starts off (324ms)
    1) toggles when pressed


  1 passing (9s)
  1 failing

  1) button to switch adapter
       toggles when pressed:
     Error: nextEventWithin timed out at within=100 ms
      at nextEventFromInternal (jdom\tester.ts:218:19)
      at runNextTicks (internal/process/task_queues.js:60:5)
      at processTimers (internal/timers.js:495:9)

```

In particular, the bus gets the event, but the device/service is wrong, so the event doesn't get delivered correctly,

**Comment by tballmsft** [comment]
@ducky64 - I think the problem may be that you are not disconnecting the devices/services spun up for a test from the bus after a test completes.

**Comment by tballmsft** [comment]
The problem appears that there are THREE devices being spun up for the second test:
```
start host
new device BG98 (54c713ce0bb2b1aa)
new device OQ86 (8c72099a0333a111)
start host
new device AP04 (b9ea0e301c924c4d)
sw device = AP04
bus event at 1000: code=1 device=BG98
```

**Comment by ducky64** [comment]
So it's expected that there's the button server and the switch server exposed by the VM, but I'm not sure what the third one is. I think I saw this on all the tests (VM or not) and it wasn't problematic, so I just assumed it was part of the bus or some other infrastructure.

**Comment by tballmsft** [comment]
Need to figure this out - I believe it's at the root of the event missing problem. I will poke at it some more.

**Comment by tballmsft** [comment]
I instrumented things, and the VMProgram is getting the down event and sending the on/off events:

```
$ node ../node_modules/mocha/bin/mocha --require ts-node/register --watch-extensions ts,tsx -p ./tsconfig.json **/buttontoswitch.spec.ts

start host
new device DW15 (2720521db993e083)
new device RE44 (4a653dc57b23b78e)
start host
new device XO48 (a6be4adb3d813b57)
SendEvent off
sw device = XO48
EXTERNAL_REQUEST button 1 down
bus event at 1000: code=1 device=DW15
SendEvent on
```

**Comment by tballmsft** [comment]
Narrowed things down... there is a problem with time.  The event is being dropped by the jdom.

**Comment by pelikhan** [comment]
Might be the triple event deboucing
________________________________

**Comment by tballmsft** [comment]
exactly... it is the event counting logic.

**Comment by tballmsft** [comment]
Updated the tracing code:

```
$ node ../node_modules/mocha/bin/mocha --require ts-node/register --watch-extensions ts,tsx -p ./tsconfig.json **/buttontoswitch.spec.ts

start host
new device FY95 (ca84849cbb40472e)
new device ON12 (399d3bc58a3cbfc1)
start host
new device SZ03 (dfca4250c04b3e9c)
SendEvent SZ03 off
sw device = SZ03
got event FY95 button 1
processEvent-start
isahead false old true missed5 true
EXTERNAL_REQUEST button 1 down
bus event at 1000: code=1 device=FY95
processEvent-end
SendEvent SZ03 on
got event SZ03 switch 1
processEvent-start
isahead true old false missed5 true

```

**Comment by tballmsft** [comment]
@ducky64 - I think I figured it out. We need to wait for the JDDevice/JDService spun up for the VM Server, just like we do for the Button.

**Comment by ducky64** [comment]
Oh, interesting. I thought by the time ROLE_BOUND gets issued (as `getRoles` waits for) the device / service would be ready to go?

File: tests/vm/buttontoswitch.spec.ts [hunk 1]
```diff
@@ -53,6 +53,7 @@ suite("button to switch adapter", () => {
 
     test("switch starts off", async () => {
         await withHarness(async (bus, button, sw) => {
+            await sw.register(SwitchReg.Active).sendGetAsync()
```
Comment (pelikhan) [inline tests/vm/buttontoswitch.spec.ts:77]: use ``refresh`` is you want to have retries as well

**Comment by tballmsft** [comment]
@ducky64 - fixed up the handling of the server.

**Comment by tballmsft** [comment]
OK. Found a nice edge case in runner. Also, we need to discuss streaming. But ready to merge.

**Pull request merged**